networkit/cpp/flow/EdmondsKarp.cpp

```diff
@@ -27,17 +26,76 @@ edgeweight EdmondsKarp::BFS(std::vector<edgeweight> &residFlow, std::vector<node
     Q.push(source);
     pred[source] = source;
     gain[source] = std::numeric_limits<edgeweight>::max();
+
     do {
         node u = Q.front();
         Q.pop();
 
         bool sinkReached = false;
         graph->forNeighborsOf(u, [&](node, node v, edgeweight weight, edgeid eid) {
-            if (((u >= v && flow[eid] < weight) || (u < v && residFlow[eid] < weight))
-                && pred[v] == none) {
+            if ((pred[v] == none)
+                && ((u > v && flow[eid] < weight) || (u < v && -flow[eid] < weight))) {
                 // only add those neighbors with rest capacity and which were not discovered yet
                 pred[v] = u;
-                gain[v] = std::min(gain[u], weight - (u >= v ? flow[eid] : residFlow[eid]));
+                auto residFlow = u > v ? flow[eid] : -flow[eid];
+                gain[v] = std::min(gain[u], weight - residFlow);
+
+                if (v != sink && !sinkReached) {
+                    Q.push(v);
+                } else {
+                    sinkReached = true;
+                }
+            }
+        });
+
+        if (sinkReached) {
+            return gain[sink];
+        }
+    } while (!Q.empty());
+
+    return 0.0;
+}
+
+edgeweight EdmondsKarp::directedBFS(const std::vector<count> &reverseEdges,
+                                    std::vector<node> &pred) const {
+    std::fill(pred.begin(), pred.end(), none);
+    pred.resize(graph->upperNodeIdBound(), none);
+    std::vector<edgeweight> gain(graph->upperNodeIdBound(), std::numeric_limits<edgeweight>::max());
```

Comment:
Since `directedBFS` might is called several times, I suggest having `gain` as a member variable of the class, so that we do not need to allocate space for it multiple times.